skiko/src/macosMain/kotlin/org/jetbrains/skiko/SkiaLayer.macos.kt

```diff
@@ -105,13 +111,60 @@ actual open class SkiaLayer() {
             }
             override fun keyDown(event: NSEvent) {
                 skikoView?.onKeyboardEvent(toSkikoEvent(event, SkikoKeyboardEventKind.DOWN))
+                keyEvent = event
+                interpretKeyEvents(listOf(event))
             }
             override fun flagsChanged(event: NSEvent) {
                 skikoView?.onKeyboardEvent(toSkikoEvent(event))
             }
             override fun keyUp(event: NSEvent) {
                 skikoView?.onKeyboardEvent(toSkikoEvent(event, SkikoKeyboardEventKind.UP))
             }
+            
+            private var keyEvent: NSEvent? = null
+            private var markedText: String = ""
+            private val kEmptyRange = NSMakeRange(NSNotFound.toULong(), 0.toULong())
+            override fun attributedSubstringForProposedRange(range: CValue<NSRange>, actualRange: NSRangePointer?) = null
+            override fun hasMarkedText(): Boolean {
+                return markedText.length > 0
+            }
+            override fun markedRange(): CValue<NSRange> {
+                if (markedText.length > 0) {
+                    return NSMakeRange(0.toULong(), (markedText.length - 1).toULong())
+                }
+                return kEmptyRange
+            }
+            override fun selectedRange() = kEmptyRange
+            override fun characterIndexForPoint(point: CValue<NSPoint>) = 0.toULong()
+            override fun doCommandBySelector(selector: COpaquePointer?) {}
+            override fun firstRectForCharacterRange(range: CValue<NSRange>, actualRange: NSRangePointer?): CValue<NSRect> {
+                val (x, y) = window.frame.useContents {
+                    origin.x to origin.y
+                }
+                return NSMakeRect(x, y, 0.0, 0.0)
+            }
+            override fun insertText(string: Any, replacementRange: CValue<NSRange>) {
+                var character = ""
+                if (string is NSAttributedString) {
+                    character = string.string
```

Comment:
Hmm, could you please explain this if and cast.

Reply:
It looks like the string could be NSString or NSAttributedString, so you need to check it first.
see apple doc: https://developer.apple.com/documentation/appkit/nstextinputclient/1438258-inserttext